lib/contextual_logger.rb

```diff
@@ -72,45 +68,52 @@ def current_context_for_thread
       Context::Handler.current_context
     end
 
-    def debug(message = nil, context = {})
-      add(Logger::Severity::DEBUG, message.nil? && block_given? ? yield : message, **context)
-    end
-
-    def info(message = nil, context = {})
-      add(Logger::Severity::INFO, message.nil? && block_given? ? yield : message, **context)
-    end
+    # In the methods generated below, we assume that presence of context means new code that is
+    # aware of ContextualLogger...and that that code never uses progname.
+    # This is important because we only get 3 args total (not including &block) passed to `add`,
+    # in order to be compatible with classic implementations like in the plain ::Logger and
+    # ActiveSupport::Logger.broadcast.
 
-    def warn(message = nil, context = {})
-      add(Logger::Severity::WARN, message.nil? && block_given? ? yield : message, **context)
-    end
+    # Note that we can't yield before `add` because `add` might skip it based on log_level. And we can't check
+    # log_level here because we might be running in ActiveSupport::Logging.broadcast which has multiple
+    # loggers, each with their own log_level.
 
-    def error(message = nil, context = {})
-      add(Logger::Severity::ERROR, message.nil? && block_given? ? yield : message, **context)
-    end
-
-    def fatal(message = nil, context = {})
-      add(Logger::Severity::FATAL, message.nil? && block_given? ? yield : message, **context)
-    end
-
-    def unknown(message = nil, context = {})
-      add(Logger::Severity::UNKNOWN, message.nil? && block_given? ? yield : message, **context)
+    LOG_LEVEL_NAMES_TO_SEVERITY.each do |method_name, log_level|
+      class_eval(<<~EOS, __FILE__, __LINE__ + 1)
+        def #{method_name}(arg = nil, **context, &block)
+          if context.empty?
+            add(#{log_level}, nil, arg, &block)
+          else
+            if arg.nil?
+              add(#{log_level}, nil, context, &block)
+            elsif block
+              add(#{log_level}, nil, context.merge(progname: arg), &block)
```

Comment:
Why did we go with the approach here of packing and unpacking the `progname` from the context when we can still pass it cleanly through `arg1`?  By doing this (un)packing, we're adding back the bug that the context is populated with unnecessary information that the formatter isn't aware of, leading to potential duplicate data being formatted by the formatter.

For example with the code as it is, if I have a formatter that looks like the following:
```ruby
-> (severity, timestamp, progname, message_with_context) do 
  message_with_context.merge(log_progname: progname).to_json
end
```
And then I call `logger.info('progname', some_context: 'hello') { 'hello world' }`, the log line will look like
```
{ 'message': 'hello world', 'progname': 'progname', 'some_context': 'hello', 'log_progname': 'progname' }
```
Is it possible to avoid this?

Reply:
> Why did we go with the approach here of packing and unpacking the progname from the context when we can still pass it cleanly through arg1

Because we can't pass it through arg1 when there's a block. (The standard `add` method only looks at the bock when arg1 is `nil`.) That's the bug that caused this ticket. I listed the constraints the comments just above.

Reply:
And we can't call the block first, because the log level needs to be checked.
And we can't check the log level because broadcast has several of them, and they hidden from us.

Reply:
Ahh, I definitely got confused by which arg is which in the original implementation. I just went back to the original method to look it up.  Can you add to the comment's what the original `add` interface is? `def add(severity, message = nil, progname = nil)`

Reply:
Done. https://github.com/Invoca/contextual_logger/pull/34/commits/822f94c5982eb245401f55e9ce12e7ade23f7e71